packages/flutter/lib/src/rendering/wrap.dart

```diff
@@ -608,7 +608,7 @@ class RenderWrap extends RenderBox with ContainerRenderObjectMixin<RenderBox, Wr
       child.layout(childConstraints, parentUsesSize: true);
       final double childMainAxisExtent = _getMainAxisExtent(child);
       final double childCrossAxisExtent = _getCrossAxisExtent(child);
-      if (runMainAxisExtent + childMainAxisExtent + spacing > mainAxisLimit) {
+      if (runMainAxisExtent + childMainAxisExtent + (childCount > 0 ? spacing : 0) > mainAxisLimit) {
         assert(childCount > 0);
```

Comment:
Prior to yesterday (#15256), this was:

```dart
if (runMainAxisExtent + childMainAxisExtent > mainAxisLimit) {
  assert(childCount > 0);
  ...
}
```

So for the `childCount == 0` case, this PR restores the previous logic.

Reply:
`childMainAxisExtent <= mainAxisLimit` is an invariant.
`runMainAxisExtent` is zero if `childCount` is zero.
This block is handling the case of needing to wrap.
If there's no children, then by definition we can't want to wrap.
If there's children, we want to wrap if placing the existing children, the spacing, then the new child, wouldn't fit.

So the clearer way to fix this would just be:

```dart
if (childCount > 0 && runMainAxisExtent + spacing + childMainAxisExtent > mainAxisLimit) { ... }
```

...except that, looking at the code, we're already adding `spacing` to `runMainAxisExtent`. So in retrospect I think Jonah's fix maybe was wrong?

Reply:
ohh!!!! i see the bug.

We're adding the spacing only if there's already a child, we should be adding the spacing unconditionally.

Reply:
that is, we should remove line 623 below, and revert line 611 to what it was like before @jonahwilliams's patch.

Reply:
hm, no, that's not right either, because we use the runMainAxisExtent later even in the 1-child case.

Reply:
ok i'm back to thinking that the clearer way to fix this is what i said it was above.
```dart
if (childCount > 0 && runMainAxisExtent + spacing + childMainAxisExtent > mainAxisLimit) { ... }
```

That is, @jason-simmons' patch is correct, but I think we should reorder the expression to more closely match what is going on.

Hopefully this time I'm not going to post this comment, look at the code again, and change my mind again.

Reply:
I see the issue - it's basically a fence-post problem right.  We could keep a separate appliedSpacing variable and start it at zero, set it to the spacing value after the first child, and then back to zero after moving to the next row/col?